llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -115,6 +115,36 @@ bool TargetLowering::parametersInCSRMatch(const MachineRegisterInfo &MRI,
   return true;
 }
 
+SDValue TargetLowering::addTokenForArgument(SDValue Chain, SelectionDAG &DAG,
+                                            MachineFrameInfo &MFI,
+                                            int ClobberedFI) const {
+  SmallVector<SDValue, 8> ArgChains;
+  int64_t FirstByte = MFI.getObjectOffset(ClobberedFI);
+  int64_t LastByte = FirstByte + MFI.getObjectSize(ClobberedFI) - 1;
+
+  // Include the original chain at the beginning of the list. When this is
+  // used by target LowerCall hooks, this helps legalize find the
+  // CALLSEQ_BEGIN node.
+  ArgChains.push_back(Chain);
+
+  // Add a chain value for each stack argument corresponding
+  for (SDNode *U : DAG.getEntryNode().getNode()->users())
```

Comment:
This is making a big assumption about the chain layout, can't you collect this when actually emitting the argument operations in the first place?

Reply:
I'm for sure not the right person to answer that question.

This code is taken from the aarch64 backend originally, which fixed its tail calls in https://github.com/llvm/llvm-project/pull/109943. From what I've been able to gather, the other backends in play worked off of the aarch64 template.

Reply:
> This code is taken from the aarch64 backend originally, which fixed its tail calls in https://github.com/llvm/llvm-project/pull/109943.

That's ARM not AArch64.

Reply:
hmm I got confused then. The `addTokenForArgument` function is from 11 years ago (and I guess has worked OK over that period):

https://github.com/llvm/llvm-project/commit/09cc56431078f38ad2b992b50f316a51da643cb0

but I know that aarch64 support for tail calls with byval arguments improved in llvm 20 (https://godbolt.org/z/b5KbTPqzY), and I guess I got that mixed up with comments in the riscv tail call implementation saying it's based on the arm code.

---

Is there a way to still share code when following this suggestion? At least as I understand it it needs some additional bookkeeping on e.g. `RISCVMachineFunctionInfo` when processing arguments in `LowerFormalArguments`?

Reply:
I'm kind of busy with some other things and a bit out of my depth here, so if someone wants to pick up this refactoring of `addTokenForArgument` that would be very helpful.